torchdata/datapipes/iter/util/plain_text_reader.py

```diff
@@ -68,6 +70,16 @@ def return_path(self, stream: Iterator[D], *, path: str) -> Iterator[Union[D, Tu
         for data in stream:
             yield path, data
 
+    def as_tuple(self, stream: Iterator[D]) -> Iterator[Union[D, Tuple]]:
+        if not self._as_tuple:
+            yield from stream
+            return
+        for data in stream:
+            if isinstance(data, list):
+                yield tuple(data)
+            else:
+                yield data
```

Comment:
In terms of performance, we might want to directly `yield tuple(data)` without checking `isinstance` every single time when `as_tuple` is specified as `True`. In that case, Users should take responsibility to handle the case.
Could you please run benchmarking using these two implementation with a long csv file?

Reply:
Hi @ejguan, I run the following benchmark:

```python
import os
import timeit
import tempfile
from torchdata.datapipes.iter import IterableWrapper, FileOpener
from torchdata.datapipes.iter.util.plain_text_reader import PlainTextReaderHelper


def impl_a(self, stream):
    if not self._as_tuple:
        yield from stream
        return
    for data in stream:
        if isinstance(data, list):
            yield tuple(data)
        else:
            yield data


def impl_b(self, stream):
    if not self._as_tuple:
        yield from stream
        return
    for data in stream:
        yield tuple(data)


fake_line = '1,2,3,4,5,6,7,8,9'
line_num = 100 * 10000

with tempfile.TemporaryDirectory() as temp_dir:
    
    temp_fpath = os.path.join(temp_dir, 'temp.csv')
    with open(temp_fpath, 'w') as f:
        f.write('\n'.join([fake_line, ] * line_num))

    datapipe1 = IterableWrapper([temp_fpath])
    datapipe2 = FileOpener(datapipe1, mode="b")
    csv_parser_dp = datapipe2.parse_csv(as_tuple=True)

    PlainTextReaderHelper.as_tuple = impl_a
    print("impl a: ", timeit.timeit(stmt=lambda: list(csv_parser_dp), number=10))

    PlainTextReaderHelper.as_tuple = impl_b
    print("impl b: ", timeit.timeit(stmt=lambda: list(csv_parser_dp), number=10))
```

Got: 

```
impl a:  13.500808166
impl b:  13.221415374999998
```

Any comments about this? ^_^

Reply:
Let's run a dummy test for the case that `not isinstance(data, list)`. To test it, you can expose `as_tuple` to `parse_csv_as_dict` and turn it on. Thank you

Reply:
Hi @ejguan , I added a benchmark for `parse_csv_as_dict`:

```python3
import csv
import os
import timeit
import tempfile
from torchdata.datapipes.iter import IterableWrapper, FileOpener
from torchdata.datapipes.iter.util.plain_text_reader import PlainTextReaderHelper
from torchdata.datapipes.iter.util.plain_text_reader import _CSVBaseParserIterDataPipe


def impl_a(self, stream):
    if not self._as_tuple:
        yield from stream
        return
    for data in stream:
        if isinstance(data, list):
            yield tuple(data)
        else:
            yield data


def impl_b(self, stream):
    if not self._as_tuple:
        yield from stream
        return
    for data in stream:
        yield tuple(data)


fake_line = '1,2,3,4,5,6,7,8,9'
line_num = 100 * 10000

with tempfile.TemporaryDirectory() as temp_dir:
    
    temp_fpath = os.path.join(temp_dir, 'temp.csv')
    with open(temp_fpath, 'w') as f:
        f.write('\n'.join([fake_line, ] * line_num))

    datapipe1 = IterableWrapper([temp_fpath])
    datapipe2 = FileOpener(datapipe1, mode="b")
    csv_parser_dp = datapipe2.parse_csv(as_tuple=True)

    PlainTextReaderHelper.as_tuple = impl_a
    print("impl a: ", timeit.timeit(stmt=lambda: list(csv_parser_dp), number=10))

    PlainTextReaderHelper.as_tuple = impl_b
    print("impl b: ", timeit.timeit(stmt=lambda: list(csv_parser_dp), number=10))

    PlainTextReaderHelper.as_tuple = impl_a
    csv_dict_parser_dp1 = _CSVBaseParserIterDataPipe(
        datapipe2, csv.DictReader, decode=True, as_tuple=True)
    csv_dict_parser_dp2 = _CSVBaseParserIterDataPipe(
        datapipe2, csv.DictReader, decode=True, as_tuple=False)
    
    print("impl a with dict as_tupe=True: ", timeit.timeit(stmt=lambda: list(csv_dict_parser_dp1), number=10))
    
    print("impl a with dict as_tupe=False: ", timeit.timeit(stmt=lambda: list(csv_dict_parser_dp2), number=10))
```

Got:
```
impl a:  13.746312875000001
impl b:  13.414058333999998
impl a with dict as_tupe=True:  25.277461583
impl a with dict as_tupe=False:  24.659475541
```

Reply:
> Let's run a dummy test for the case that `not isinstance(data, list)`. To test it, you can expose `as_tuple` to `parse_csv_as_dict` and turn it on. Thank you

Do you mean we should add the as_tuple argument to `CSVDictParserIterDataPipe`?

Reply:
How about `impl b` with dict?

Reply:
Hi, there are the result of impl_b with dict parser:

```
impl a:  13.375169375
impl b:  13.102719625
impl a with dict as_tupe=True: 24.521925125000003
impl a with dict as_tupe=False: 23.991537542000003
impl b with dict as_tupe=True: 24.552794458999998
impl b with dict as_tupe=False: 24.026250667
```